src/borg/archive.py

```diff
@@ -1298,29 +1235,32 @@ def __init__(self, *, metadata_collector, cache, key,
     def create_helper(self, path, st, status=None, hardlinkable=True):
         safe_path = make_path_safe(path)
         item = Item(path=safe_path)
-        hardlink_master = False
         hardlinked = hardlinkable and st.st_nlink > 1
+        update_map = False
         if hardlinked:
-            source = self.hard_links.get((st.st_ino, st.st_dev))
-            if source is not None:
-                item.source = source
-                status = 'h'  # hardlink (to already seen inodes)
-            else:
-                hardlink_master = True
-        yield item, status, hardlinked, hardlink_master
-        # if we get here, "with"-block worked ok without error/exception, the item was processed ok...
+            status = 'h'  # hardlink
+            hlid, chunks = self.hlm.retrieve(id=(st.st_ino, st.st_dev), default=(None, None))
+            if hlid is None:
+                update_map = True
+                hlid = hardlink_id(item._dict['path'])
+            item.hlid = hlid
+            if chunks is not None:
+                item.chunks = chunks
+        yield item, status, hardlinked
         self.add_item(item, stats=self.stats)
-        # ... and added to the archive, so we can remember it to refer to it later in the archive:
-        if hardlink_master:
-            self.hard_links[(st.st_ino, st.st_dev)] = safe_path
+        if update_map:
+            # remember the hlid of this fs object and if the item has chunks,
+            # also remember them, so we do not have to re-chunk a hardlink.
+            chunks = item.chunks if 'chunks' in item else None
+            self.hlm.remember(id=(st.st_ino, st.st_dev), info=(hlid, chunks))
```

Comment:
instead of computing the `hlid = hardlink_id(path)` (see 1245), we could also compute it from `hardlink_id(st_ino, st_dev)`.

if we encounter the next related hardlink, it will have the same `(st_ino, st_dev)` (we get that from `stat`, together with `st_nlinks >= 1`). so we can compute the `hlid` from that and do not need to remember it, but only the chunks list.

Reply:
doing that...